daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs

```diff
@@ -267,6 +264,42 @@ runStart
             case mbJsonApiPort of
                 Nothing -> f sandboxPh
                 Just jsonApiPort -> withJsonApi sandboxPort jsonApiPort args f
+        doCodegen projectConfig =
+          forM_ [minBound :: Lang .. maxBound :: Lang] $ \lang -> do
+            mbOutputPath :: Maybe String <-
+              requiredE ("Failed to parse codegen entry for " <> showLang lang) $
+              queryProjectConfig
+                ["codegen", showLang lang, "output-directory"]
+                projectConfig
+            whenJust mbOutputPath $ \_outputPath -> do
+              runCodegen lang []
+        doUploadDar =
+          runLedgerUploadDar (LedgerFlags Nothing Nothing Nothing Nothing 10) Nothing
+        listenForKeyPress projectConfig = do
+          hSetBuffering stdin NoBuffering
+          void $
+            forkIO $
+             do
+              threadDelay 20000000
+              -- give sandbox 20 secs to startup before printing rebuild instructions
+              forever $ do
+                printRebuildInstructions
+                c <- getChar
+                when (c `elem` ("rR" :: String)) $ rebuild projectConfig
```

Comment:
The supposedly more compact code ended up longer because of string overloading. Funny.
```suggestion
                when (c == 'r' || c == 'R') $ rebuild projectConfig
```